### PR TITLE
Stabilize `count`, `ignore`, `index`, and `len` (`macro_metavar_expr`)

### PR DESCRIPTION
**Pull request by c410-f3r**
# Stabilization proposal

This PR proposes the stabilization of a subset of `#![feature(macro_metavar_expr)]` or more specifically, the stabilization of `count`, `ignore`, `index` and `len`.

Tracking issue: #83527
Version: 1.80 (June 13 2024 => beta, July 25 2024 => stable).

### Count

The number of times a meta variable repeats in total.

The output of `count`  depends on where it is placed as well the provided index. If no index is provided, then it will always start at the innermost level.

```rust
macro_rules! count_idents {
    ( $( $i:ident ),* ) => {
        ${count($i)}
    };
}

fn main() {
    assert_eq!(count_idents!(a, b, c), 3);
}
```

### Ignore

Binds a meta variable for repetition, but expands to nothing.

```rust
macro_rules! count {
    ( $( $i:stmt ),* ) => {{
        0 $( + 1 ${ignore($i)} )*
    }};
}

fn main() {
    assert_eq!(count!(if true {} else {}, let _: () = (), || false), 3);
}
```

### Index

The current index of the inner-most repetition.

`index` is a counter that starts from 0 until the end of the repetition.

```rust
trait Foo {
    fn bar(&self) -> usize;
}

macro_rules! impl_tuple {
    ( $( $name:ident ),* ) => {
        impl<$( $name, )*> Foo for ($( $name, )*)
        where
            $( $name: AsRef<[u8]>, )*
        {
            fn bar(&self) -> usize {
                let mut sum: usize = 0;
                $({
                    const $name: () = ();
                    sum = sum.wrapping_add(self.${index()}.as_ref().len());
                })*
                sum
            }
        }
    };
}

impl_tuple!(A, B, C, D);

fn main() {
}
```

### len

The current index starting from the inner-most repetition.

`len` indicates the sum of meta variable elements, aka length, of the current repetition.

```rust
macro_rules! array_3d {
    ( $( $( $number:literal ),* );* ) => {
        [
            $(
                [
                    $( $number + ${len()}, )*
                ],
            )*
        ]
    };
}

fn main() {
    assert_eq!(array_3d!(0, 1; 2, 3; 4, 5), [[2, 3], [4, 5], [6, 7]]);
}
```

## Motivation

Meta variable expressions not only facilitate the use of macros but also allow things that can't be done today like in the `$index` example.

An initial effort to stabilize this feature was made in https://github.com/rust-lang/rust/pull/111908 but ultimately reverted because of possible obstacles related to syntax and expansion.

Nevertheless, https://github.com/rust-lang/rust/issues/83527#issuecomment-1744822345 tried to address some questions and fortunately the lang team accept https://github.com/rust-lang/rust/pull/117050 the unblocking suggestions.

Here we are today after ~4 months so everything should be mature enough for wider use.

## After RFC changes

In order to unblock progress, some changes were applied in https://github.com/rust-lang/rust/pull/118958. These changes did not trigger any opposition debates and were quickly integrated.

1. The depth direction of `count` changed from `outermost-to-innermost` to `innermost-to-outermost`.
2. Metavariables were originally referenced by their names like in `count(some_metavariable)` but now they must be preceded by `$`. For example, `count($some_metavariable)`.

## What isn't stabilized

`$$` is not being stabilized due to unresolved concerns.

## History

* 2021-02-22, [RFC: Declarative macro metavariable expressions](https://github.com/rust-lang/rfcs/pull/3086)
* 2021-03-26, [Tracking Issue for RFC 3086: macro metavariable expressions](https://github.com/rust-lang/rust/issues/83527)
* 2022-02-01, [Implement macro meta-variable expressions](https://github.com/rust-lang/rust/pull/93545)
* 2022-02-25, [[1/2] Implement macro meta-variable expressions](https://github.com/rust-lang/rust/pull/94368)
* 2022-03-11, [[2/2] Implement macro meta-variable expressions](https://github.com/rust-lang/rust/pull/94833)
* 2022-03-12, [Fix remaining meta-variable expression TODOs](https://github.com/rust-lang/rust/pull/94884)
* 2019-03-21, [[macro-metavar-expr] Fix generated tokens hygiene](https://github.com/rust-lang/rust/pull/95188)
* 2022-04-07, [Kickstart the inner usage of macro_metavar_expr](https://github.com/rust-lang/rust/pull/95761)
* 2022-04-07, [[macro_metavar_expr] Add tests to ensure the feature requirement](https://github.com/rust-lang/rust/pull/95764)
* 2023-05-24, [[RFC-3086] Restrict the parsing of `count`](https://github.com/rust-lang/rust/pull/111908)
* 2023-10-22, [[RFC 3086] Attempt to try to resolve blocking concerns](https://github.com/rust-lang/rust/pull/117050)

## Tests

This list is a subset of https://github.com/rust-lang/rust/tree/master/tests/ui/macros/rfc-3086-metavar-expr.

* [Ensures that nested macros have correct behavior](https://github.com/rust-lang/rust/blob/master/src/test/ui/macros/rfc-3086-metavar-expr/dollar-dollar-has-correct-behavior.rs)

* [Compares produced tokens to assert expected outputs](https://github.com/rust-lang/rust/blob/master/src/test/ui/macros/rfc-3086-metavar-expr/feature-gate-macro_metavar_expr.rs)

* [Checks the declarations of the features](https://github.com/rust-lang/rust/blob/master/src/test/ui/macros/rfc-3086-metavar-expr/required-features.rs)

* [Verifies all possible errors that can occur due to incorrect user input](https://github.com/rust-lang/rust/blob/master/src/test/ui/macros/rfc-3086-metavar-expr/syntax-errors.rs)

## Possible future work

With enough consensus, other nightly expressions can be added for experimentation and possibly stabilized in the future. For example, https://github.com/rust-lang/rust/pull/118958.

Thanks @markbt for creating the RFC and thanks to @petrochenkov and @mark-i-m for reviewing the implementations.

### CODE REVIEW

**Comment by fmease** [comment]
@c410-f3r Has this been documented in the Reference already? The box is not ticked over at the tracking issue.

**Comment by joshtriplett** [comment]
@rfcbot merge

@rfcbot concern document-in-reference

**Comment by joshtriplett** [comment]
@rfcbot cancel

**Comment by joshtriplett** [comment]
@rfcbot merge

@rfcbot concern document-in-reference

**Comment by jhpratt** [comment]
cc @rust-lang/wg-macros

**Comment by RalfJung** [comment]
> Length
> 
> The current index starting from the inner-most repetition.

What does this mean, and why does "length" not show up in the explanation of the `length` "function"? I would expect this to be the length of the inner-most repetition, is that not the case?

All these "functions" expand to integers -- of which type? Or do they end up being like integer literals without a suffix, i.e., the concrete integer type gets inferred following the usual rules?

**Comment by Veykril** [comment]
Note that the tracking issue still has an unresolved question/bug open https://github.com/rust-lang/rust/issues/83527

> [Figure out problems around hygiene](https://github.com/rust-lang/rust/issues/83527#issuecomment-1073506711)

Has this been resolved?

**Comment by joshtriplett** [comment]
@rfcbot concern hygiene-handled?

**Comment by Veykril** [comment]
Ah actually, it seems the checkbox was never ticked? https://github.com/rust-lang/rust/pull/95188

**Comment by joshtriplett** [comment]
@rfcbot resolved hygiene-handled?

**Comment by lcnr** [comment]
what is the difference between `index` and `length`?

**Comment by tgross35** [comment]
Are the functions that take depth such as `${count(ident, depth)}` included as part of this?

Nonblocker, diagnostics are pretty bare bones with no error codes or help

```
error: `count` can not be placed inside the inner-most repetition
 --> src/main.rs:7:19
  |
7 |             dbg!(${count($i, 0)});
  |                   ^^^^^^^^^^^^^^
```

**Comment by slanterns** [comment]
> * 2019-03-21, [[macro-metavar-expr] Fix generated tokens hygiene](https://github.com/rust-lang/rust/pull/95188)

Date seems incorrect :)

**Comment by nagisa** [comment]
To follow up on the question on the depth of `count`, does "in total" mean this considers all levels of repetition? Like, if I write`$( $a: ident: $($b: literal),* );+` and this matches a `a: 1, 2, 3; b: 4, 5, 6`, do I get `count($b)` of 3 twice, or a `count($b)` of 6 twice? Or does it depend on where I write the `count` at?

The RFC text does not really make this clear…

**Comment by c410-f3r** [comment]
@fmease
> Has this been documented in the Reference already?

No but documentation will be provided as soon as possible.

@RalfJung

> What does this mean, and why does "length" not show up in the explanation of the length "function"? I would expect this to be the length of the inner-most repetition, is that not the case?

`length` indicates  the sum of meta variable elements, aka length, of the **current** repetition and that is why, contrary the `count`, it can only be placed inside repetitions.

The nested repetitions of https://user-images.githubusercontent.com/17877264/158820060-035b2548-9ea9-4cbb-850b-e5836cf892fe.png illustrate better this concept.

More information is available at https://rust-lang.github.io/rfcs/3086-macro-metavar-expr.html#index-and-length

> All these "functions" expand to integers -- of which type? Or do they end up being like integer literals without a suffix, i.e., the concrete integer type gets inferred following the usual rules?

`length` and `count` are expanded to unsuffixed integer literals.

More information is available at https://rust-lang.github.io/rfcs/3086-macro-metavar-expr.html#count

https://github.com/rust-lang/rust/blob/df8ac8f1d74cffb96a93ae702d16e224f5b9ee8c/compiler/rustc_expand/src/mbe/transcribe.rs#L677-L683

@lcnr 

> what is the difference between index and length?

`length` indicates  the sum of meta variable elements of the **current** repetition while `index` is a counter that starts from 0 until the end of the repetition.

```rust
macro_rules! index {
    ( $( $tt:tt ),* ) => {
        [$( ${ignore($tt)} ${index()}, )*]
    };
}

macro_rules! length {
    ( $( $tt:tt ),* ) => {
        [$( ${ignore($tt)} ${length()}, )*]
    };
}

fn main() {
    // 3 tokens generate an increment list of 3 elements starting from 0
    assert_eq!(index!(A, B, C), [0, 1, 2]);
    
    // 4 tokens generate a list of 4 elements containing 4
    assert_eq!(length!(A, B, C, D), [4, 4, 4, 4]);
}
```

More information is available at https://rust-lang.github.io/rfcs/3086-macro-metavar-expr.html#index-and-length

@ slanterns 
> Date seems incorrect :)

Thanks

@nagisa

> To follow up on the question on the depth of count, does "in total" mean this considers all levels of repetition? Like, if I write $( $a: ident: $($b: literal),* );+ and this matches a a: 1, 2, 3; b: 4, 5, 6, do I get count($b) of 3 twice, or a count($b) of 6 twice? Or does it depend on where I write the count at?

`count` depends on where it is placed as well the provided index. If no index is provided, then it will always start at the innermost level.

So in your example a declaration of `$( $a:ident: $($b: literal),* );+`  called with `a: 1, 2, 3; b: 4, 5, 6,` will produce `3, 3` if called inside the outermost repetition or `6`  if called without any repetition.

```rust
// Other examples

macro_rules! no_repetition {
    ( $( $a:ident: $( $b:literal ),* );+ ) => {
        [${count($b)}]
    };
}

macro_rules! no_repetition_with_index {
    ( $( $a:ident: $( $b:literal ),* );+ ) => {
        [${count($b, 1)}]
    };
}

macro_rules! outermost_repetition {
    ( $( $a:ident: $( $b:literal ),* );+ ) => {
        [$( ${ignore($a)} ${count($b)}, )+]
    };
}

fn main() {
    // 1 2 3 4 5 = 5 elements
    assert_eq!(no_repetition!(a: 1, 2, 3; b: 4, 5), [5]);
    
    // a b = 2 elements
    assert_eq!(no_repetition_with_index!(a: 1, 2, 3; b: 4, 5), [2]);
    
    // 1 2 3 = 3 elements, 4 5 = 2 elements
    assert_eq!(outermost_repetition!(a: 1, 2, 3; b: 4, 5), [3, 2]);
}
```

https://user-images.githubusercontent.com/17877264/158820060-035b2548-9ea9-4cbb-850b-e5836cf892fe.png might help you but it was created in a time when `count` started at the outermost level.

##### Documentation updated

**Comment by jdahlstrom** [comment]
So does `${count()}` equal `${length()}` in the most common case of having only a single level of repetition?

**Comment by danielhenrymantilla** [comment]
@jdahlstrom 
IIUC, I believe:
```rs
[$( $x / ${length()} ),*]
```
to be morally equivalent to:
```rs
let n = ${count(x)};
[$( $x / n ),*]
```

**Comment by tgross35** [comment]
Would it be possible to prepare the PR for the reference before this lands? It seems like there is quite a bit of uncertainty about how all the features work, which makes me thing that perhaps this feature isn't very widely tested. As semantics have changed a few times since the RFC, that isn't the best reference either.

Quick search for usage that isn't in rustc forks https://github.com/search?q=lang%3Arust+%2F%28%3F-i%29%5C%24%5C%7B%28ignore%7Ccount%7Clength%7Cindex%29%5C%28.*%5C%7D%2F+-path%3Acompiler%2F+-path%3Alibrary%2F+-path%3Asrc%2Ftools%2Frust-analyzer&type=code

**Comment by c410-f3r** [comment]
> > Has this been documented in the Reference already?

> No but documentation will be provided as soon as possible.

As commented above, reference will be provided as soon as possible.

Semantic changed only once when `count` moved from outermost to innermost. https://github.com/rust-lang/rust/pull/117050 also required the addition of `$` but this is more syntax-related than an actual change of how the feature works, everything else remains the same as when the RFC was approved 2 years ago.

**Comment by workingjubilee** [comment]
Is there any realistic pattern to suffix these literal expansions or nah?

**Comment by programmerjake** [comment]
> Is there any realistic pattern to suffix the integer literal expansions or nah?

well, I expect `${concat(${length()}, _u32)}` to work at some point...

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by jdahlstrom** [comment]
I brought this up in the doc PR but it belongs here – `length` should probably be renamed `len`. The latter is de facto standard in the standard library, whereas the former is only used in a single unstable API. These metafunctions aren’t library items, but should presumably still be consistent with established names.

**Comment by workingjubilee** [comment]
I'm slightly worried about random inference regressions punishing this particularly hard if there's no standard/easy way to proof the emitted literals against such. I guess we can use a block expression for each one?

**Comment by joshtriplett** [comment]
@rfcbot resolve document-in-reference

**Comment by workingjubilee** [comment]
> I'm slightly worried about random inference regressions punishing this particularly hard if there's no standard/easy way to proof the emitted literals against such. I guess we can use a block expression for each one?

I've since-then thought of a few more alternatives to both "expand to code with type annotations" _and_ composing with the proposed `${concat}` metavariable expression, so I'm more confident that this will be okay eventually, even if it does prove to be a practical concern.

**Comment by workingjubilee** [comment]
> I brought this up in the doc PR but it belongs here – `length` should probably be renamed `len` before stabilization. The latter is de facto standard in the standard library, whereas the former is only used in a single unstable API. These metafunctions aren’t library items of course, but should presumably still be consistent with established names.

for everyone following along, @jdahlstrom's proposal was yote before the relevant teams, accepted, and merged in #124987.

**Comment by tgross35** [comment]
After looking into this a bit deeper, I think some of the reason this API feels so confusing is that we do not have a way to refer to specific repetition groups directly. Instead, we try to refer to specific groups based on the variables they capture and an offset - this gets messy. E.g., this example from the reference PR:

```rust
macro_rules! innermost1 {
    ( $( $a:ident: $( $b:literal ),* );+ ) => {
        [$( $( ${ignore($b)} ${index(1)}, )* )+]
    };
}
```

Here you have to count how many repetition groups exist in the match expression, then count how many repetition groups the `index(1)` call is nested in, then backtrack by one to figure out what exactly is getting indexed. This is a lot to keep track of mentally.

I posted a [pre-RFC](https://internals.rust-lang.org/t/pre-rfc-named-capture-groups-for-macros/20883) to discuss the idea of being able to name capture groups, which could then be referred to directly within metavar expressions. The above would become:

```rust
macro_rules! innermost1 {
    ( $outer_rep:( $a:ident: $inner_rep:( $b:literal ),* );+ ) => {
        [$outer_rep( $inner_rep( ${ignore($b)} ${index($outer_rep)}, )* )+]
    };
}
```

It is significantly easier to immediately see that `index($outer_rep)` refers to the outermost repetition group.

`len($var)` also tries to use `$var` as a proxy for the group it captures. `len($outer_rep)` (or possibly better `count_repetitions($outer_rep)`) would be more accurate here.

A nice side effect of the proposal is that our diagnostics could become more expressive, rather than gesturing at some code and trying to explain repetition errors in terms of what is captured.

One other thing is that the term `depth` seems backwards, since increasing the `depth` value actually _decreases_ the level of depth/nesting. `parent` seems like a slightly better term for this argument, there might be something better yet. This can of course be fixed at any time without breaking anything.

In short: I feel great about stabilizing `index()` (no `depth`) and `ignore($var)` now. However, it seems like the API for `len` and `count` could be improved significantly if they were given a way to refer to groups themselves, rather than indirectly through captures.

(adding named repetition groups and using them for `count`/`len` would be backward compatible, it would just be nice to not support the offset/depth-based API at all.)

**Comment by programmerjake** [comment]
so then `count` would be something like the following?

```rust
macro_rules! m {
    ($outer:(a $middle:(b $inner:(c $v:ident)*)*)*) => {
        println!("{}", stringify!(
            ${count($outer, $inner)}
            [$outer(${count($middle, $inner)})*]
            ${count($outer, $middle)}
            ${count($outer)}
            [$outer(${count($middle)})*]
            [$outer([$middle(${count($inner)})*])*]
        ));
    }
}
m!(a b c v0 c v1 a a b b);
// prints (without the comments):
// 2 // two inners inside outer
// [ 2 0 0 ] // two inners inside the first middle, zero for the rest
// 3 // three middles inside outer
// 3 // three repetitions inside outer
// [ 2 0 2 ] // two repetitions inside first middle, zero for second, two for third
// // two repetitions inside inner in first middle, no inners in second middle, no repetitions inside both inners in third middle
// [ [ 2 ] [ ] [ 0 0 ] ]
```

**Comment by tgross35** [comment]
Those all have the results I would expect, except I think `[$outer(${count($middle)})*]` would be `[1 0 2]` since there is only one `$middle` within the first `$outer`. And for the first three, I wasn't thinking multiple arguments would be needed since the variable is already scoped to a repetition group.

I am loosely using these rules:
- When querying more deeply nested variables/groups (downward), expressions only "see" those instances that are direct descendants
- When querying equally or less deeply nested variables/groups (upward), expressions "see" everything (i.e. neighbors, parents, everything at the same level)

```text
m!(a b c v0 c v1 a a b b);

              /--------- (root) macro ------------\
             /                |                    \
        (outer0) a         (outer1) a           (outer2) a
           |                                   /        \
       (middle0) b                      (middle1) b    (middle2) b
      /        \
(inner0) c  (inner1) c
    |           |
 (v0) v0      (v1) v1
```

So, for your first three examples where I think I would change the matcher:

1. `${count($outer, $inner)}` would just be `${count($inner)}` -> `2`. It gets called once at the root level and sees all instances of `$inner` that could possibly exist (2 here)
2. `[$outer(${count($middle, $inner)})*]` would just be `[$outer(${count($inner)})*]` -> `[2 0 0]`. It gets called once at `outer0` (sees two `$inner`s as descendants), once at `outer1` (no `$inners`), and once at `outer2` (no `$inners`)
3. `${count($outer, $middle)}` would just be `${count($middle)}` -> `3`. It gets called once at the root level and sees all instances of `$middle` as descendants to count

Explanation of the remaining three:

4. `${count($outer)}` -> 3: root scope, sees 3 total repetitions of `$outer`, returns 3
5. `[$outer(${count($middle)})*]` -> `[1 0 2]` sees one `middle` as a child at `outer0`, none at `outer1`, 2 at `outer2`.
6. `[$outer([$middle(${count($inner)})*])*]` -> `[ [ 2 ] [ ] [ 0 0 ] ]` just repeating the above

And then some others possibilities:

7. `[$outer([$middle(${count($v})]` -> `[[ 2 ] [ ] [ 0 0 ]]`: regular metavars can still be counted. In this case `count($v)` and `count($inner)` happen to give the same results
8. `[$outer( ${count($outer)} )]` -> `[3 3 3]`: Instantiated at `outer1`, `outer2`, `outer3`, can see everything at that level and above. `${len()}` would basically be an alias of this.
9. `[$outer([$middle([$inner( ${count($outer)} )*])*])*]` -> `[[[3 3] [ ] [ ]]`: similar to above, just deeper nesting

(Examples 8 and 9 show cases where the current `len()` and `count()` produce identical results, with the only difference being restrictions on where they can be used. If we got named capture groups I think we could likely combine these, or at least make their differences more clear).

**Comment by c410-f3r** [comment]
The use of a named placeholder to indicate the depth of a repetition seems orthogonal to the stabilization of the current approach accepted in the RFC. As such, it probably shouldn't be a blocking concern.

On the merits, the suggested approach looks overly verbose but nothing prevents a future inclusion if so desired by the responsible team.

**Comment by RalfJung** [comment]
I tend to agree with @tgross35 -- having to refer to capture groups by index is a really awkward and fragile API, I don't think we should stabilize that if better alternatives exist. After all we generally let the user name things they need to refer to (variables, functions, types) instead of using opaque numbers and counting.

**Comment by workingjubilee** [comment]
Is there consensus for `${ignore($var)}` and `${count($var)}` at least? Those would make certain patterns much easier to handle.

**Comment by RalfJung** [comment]
`ignore` seems like a no-brainer to me. (Not sure this is the best possible name, but I can't come up with a better one either.)

`count` does seem to have very surprising interaction with repetitions. Judging from the [reference PR](https://github.com/rust-lang/reference/pull/1485):
```rust
macro_rules! count_value_nested {
    ( $( $name:ident: $( $value:literal ),* );+ ) => {
        [ $(
            // using `count` within a repetition group will return the number
            // of times `$value` is matched _within that group_.
            ${count($value)},
        )+ ]
    };
}
```
So, this is a repetition group that only contains `$value`. And somehow that becomes a once-per-`$name` repetition? That's not how repetition behaves anywhere else, is it? I would have expected that I must use `$name` somewhere in this repetition group to make this a once-per-`$name` group, and then I can count the `$value` in it -- the argument of `$count` would not be part of the automatic detection of the repetition group in my mental model, that has to already be clear without.

Which repetition group does this count?

```rust
macro_rules! count_test {
    ( $(supername:ident { $( $name:ident: $( $value:literal ),* );+ } )* ) => {
        [ $(
            ${count($value)},
        )+ ]
    };
}
```

This could reasonably be `for i in supername { i.total_numer_of_value() }` but also `for i in supername { for j in i.name { j.total_numer_of_value() } }`.

**Comment by RalfJung** [comment]
[Trying this out](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f5ecced87580bd2bd5a8a4645cc1c983), it seems like like $(...)` ends up iterating over the `supername`s, not the `name`s.

Usually a metavar "must appear in exactly the same number, kind, and nesting order of repetitions in the transcriber as it did in the matcher". With `count` that seems to be different. Is it something like "must appear in *strictly fewer* repetitions than it did in the matcher, and then it counts the number of inner repetitions in the current instance of the outer repetitions it occurs in"?

**Comment by workingjubilee** [comment]
looks kinda like `group.iter().flat_map(|expansion| expansion.get(metavar)).count()`

**Comment by c410-f3r** [comment]
As previously commented, the suggestion around named placeholders is orthogonal to the current behavior approved in the RFC and both can co-exist in the future.

Fate imposed yet another ***syntax-related*** blocking concern in a stabilization PR, as such, I ask the responsible party to decide the final outcome.

@rustbot labels +I-lang-nominated

**Comment by tmandry** [comment]
@rfcbot concern depth levels are confusing

Before reading this thread I had a strong reaction to the RFC exactly in line with @RalfJung's, and I had thoughts along the lines of what @tgross35 is proposing. So I think there's something worth paying attention to here. I see it as a major downside, quite possibly a failure, of the lang team process that we got this far without such syntax concerns being surfaced. With that said, even though the concern is a late breaking one, it feels significant enough to me that I don't think it should be minimized.

Rust's declarative macro syntax can already be quite intimidating to the uninitiated, and we should be careful in how we spend our "readability budget" here. This emphasis on _readability_ is the reason I'm not that swayed by the argument of future compatibility: Even if we add named expansions later, the syntax for unnamed depth levels will always exist if we stabilize it now, and people who read Rust macro code will have to know about it.

To make my concern more concrete: I think adding recursion depths is a significant step up in both _capability_ and _complexity_, but I would want that to be equally matched with stronger _usability_ as well as _motivation_ for the feature. The usability piece is quite lacking, in my opinion, in that the recursion depths are unlabeled integers without an obvious meaning. I certainly think they are non-obvious to the reader, and in some cases even to the writer who already knows about this feature (I speculate that many users might forget how `count` works with nested metavars and depths, even after using it a few times, because it doesn't seem to work like anything else in the language). I also don't see a motivation for these specific capabilities laid out in the RFC or the stabilization report.

---

In the interest of making forward progress, I'll put forward a proposal that I frame as a question. The implied question in each of these is how much it limits the use cases enabled by these features. If the answer is "not that much" I think we should go forward with stabilizing this subset for now; otherwise, I hope it's a useful starting point for getting things moving.

* Can the stabilization proposal be subsetted not to support depth levels on any of these operators?
* Can `count` specifically be limited to supporting only metavars with a single recursion level?
    * If that's too limiting: What is the motivation for counting _all_ nested expansions instead of only counting the outermost level (the one that would be expanded in the next use of `$()*`)? If hypothetically we switched it to instead only count the outermost level (without supporting explicit depths but holding open the possibility to support them in the future), would that enable significantly more use cases?

**Comment by RalfJung** [comment]
> If that's too limiting: What is the motivation for counting all nested expansions instead of only counting the outermost level (the one that would be expanded in the next use of $()*)?

I don't exactly know what you mean here, could you give an example?

FWIW, I experimented with `count` a bit and [here](https://github.com/rust-lang/reference/pull/1485#discussion_r1620054777) are the results. Without named repetitions, I don't think there is another interpretation that makes sense. Possibly there is a subset that makes sense, like only supporting `count($x)` when `x` is exactly one level inside the current nesting -- but it's not clear that that actually makes it more understandable.

**Comment by tmandry** [comment]
> I don't exactly know what you mean here, could you give an example?

What I meant is, why not make `count($x)` equivalent to `count($x, 1)` instead of `count($x, ∞)`? We would then have to invent syntax for the latter if we wanted to preserve expressiveness, but this default aligns more with my expectations.

**Comment by RalfJung** [comment]
AFAIK it is currently equivalent to count($x, 0), which does align with my intuition once I re-familiarized myself with how nested capture groups work. I don't know what infinite depth would even mean.

**Comment by tmandry** [comment]
I'm not sure what count(x, 0) would mean, based on this from the RFC:

> If repetitions are nested, then an optional depth parameter can be used to limit the number of nested repetitions that are counted. For example, a macro expansion like:
> 
>     ${count(x, 1)} ${count(x, 2)} ${count(x, 3)} $( a $( b $( $x )* )* )*
>
> The three values this expands to are the number of outer-most repetitions (the number of times a would be generated), the sum of the number of middle repetitions (the number of times b would be generated), and the total number of repetitions of $x.

Reading this over again though, I no longer think it makes sense to limit to 1 by default. `${count(x)}` should of course count the number of expansions for `$x`, not the number of expansions for the next-expandable metavariable that `$x` happens to be nested in. We could be conservative in only allowing it in the same contexts where `$x` can already be expanded, but I can see it being useful outside of those contexts.

I still think specifying a max depth level is non-obvious and of dubious value, since you could always rewrite the excerpt above as something like the following; are there cases where you can't do this?

     ${count(a)} ${count(b)} ${count(x)} $( a $( b $( $x )* )* )*

**Comment by RalfJung** [comment]
> I'm not sure what count(x, 0) would mean, based on this from the RFC:

I have linked this already have, see [here](https://github.com/rust-lang/reference/pull/1485#discussion_r1620054777). That's what I reverse engineered based on my experiments, anyway. The default is definitely 0, according to that PR.

Maybe the semantics changed since the RFC was accepted?

> I still think specifying a max depth level is non-obvious and of dubious value, since you could always rewrite the excerpt above as something like the following; are there cases where you can't do this?

I think the issue is around matchers like `$( [[ $(  {{ $(  $x, )* }} )* ]] )*` where there is no variable in the outer repetition groups that one can refer to.

**Comment by c410-f3r** [comment]
I will try to address all questions this weekend in my free time.

cc https://github.com/rust-lang/reference/pull/1485#issuecomment-2182766692

**Comment by c410-f3r** [comment]
> I see it as a major downside, quite possibly a failure, of the lang team process that we got this far without such syntax concerns being surfaced.

Thank you. `macro_metavar_expr` has been around for +3 years and its implementation has been available for +2 years. Although part of the game, it is surprising and demotivating to see syntax concerns in stabilizing PRs (https://github.com/rust-lang/rust/pull/95860).

> The only point that gives me pause around count is -- if/when we have named repetition groups, how will they work with count, and is that compatible with today's behavior?

It is more complex but not impossible IMO. One way is to attach a corresponding matcher depth to each named repetition and then evaluate which ones can be associated based on the current transcriber position.

Syntax-wise, the distinction between numbers and strings can create appropriated implementation branches.

> What I meant is, why not make count($x) equivalent to count($x, 1) instead of count($x, ∞)? We would then have to invent syntax for the latter if we wanted to preserve expressiveness, but this default aligns more with my expectations.

> AFAIK it is currently equivalent to count($x, 0),

Before https://github.com/rust-lang/rust/pull/117050 `count($x)` was equal to `count($x, N)` with `N` being the number of nested repetitions minus one. Now `count($x)` is equal to `count($x, 0)` which can also be seen as the innermost repetition.

https://github.com/rust-lang/reference/pull/1485#issuecomment-2182766692 will probably provide a better overview.

> I still think specifying a max depth level is non-obvious and of dubious value, since you could always rewrite the excerpt above as something like the following; are there cases where you can't do this?

>  ```${count(a)} ${count(b)} ${count(x)} $( a $( b $( $x )* )* )*```

Unfortunately it is not possible to count without a metavariable. AFAICT, `a` and `b` are just references used in the RFC to illustrative how the algorithm works.

But yeah, the specification of a depth is non-obvious as it depends on the number of nested matcher repetitions as well as the positioning.

> the people involved in the implementation

@markbt, @petrochenkov and @mark-i-m

**Comment by tmandry** [comment]
I see, I was confused by reading the RFC for more detail, but that is now outdated. @c410-f3r Can you please include a list of differences from the RFC in the stabilization report (with rationale; feel free to summarize and link to more context)?

**Comment by c410-f3r** [comment]
> I see, I was confused by reading the RFC for more detail, but that is now outdated. @c410-f3r Can you please include a list of differences from the RFC in the stabilization report (with rationale; feel free to summarize and link to more context)?

Sure. The report has been updated.

**Comment by alex-semenyuk** [comment]
@c410-f3r
From triage team. Does it ready for review? if so could you please fix merge conflicts before

**Comment by c410-f3r** [comment]
Well, I can rebase but I don't think it will make much of a difference. The actual issue lies within the decision around the use of indices in `count`, `index` and `len`.

This PR has been waiting for more than 2 months for a response from the responsible team. AFAICT, all questions posted here as well as in https://github.com/rust-lang/reference/pull/148 have already been answered to the best of my ability.

**Comment by fmease** [comment]
What's the context? Why was this closed?

**Comment by Lokathor** [comment]
Since the person who closed it is the author, who in September had already been waiting for "more than 2 months", and it's been many months since, and it's even been a month since it was Lang nominated with no update from T-lang, I have to assume they lost interest in going any further with this.

**Comment by ZaneErebos** [comment]
I really hope this does not get abandoned. I was tracking this for a while and it would be a great addition. In any case, thank you to the author for all their hard work and patience

**Comment by jhpratt** [comment]
I will take this over (whether as this PR or a new one) if the author is no longer interested.
